### DISCUSSION TITLE
When should we use larger encoder/decoder architectures?

### DISCUSSION THREAD

**Discussion by cbeck22** [zhonge/cryodrgn#400] [Q&A] [answered]
Hi!

Upon reviewing the methods section of some papers that have used cryoDRGN for their datasets, I've noticed that some papers have used larger encoder/decoder architectures than the recommended 1024x3 architectures. I've also found one paper that used different architectures for the encoder and decoder, i.e. 2048x3 for the encoder and 1024x3 for the decoder. Is there any reason to use larger architectures? Are larger architectures able to characterize more complex motions? Is there any method to separate subtler motions into more distinct clusters?

Cheers,
cbeck

**Comment by michal-g** [comment, marked as answer]
Hi, to be brief, yes, it is generally advantageous to use larger architectures if possible, the intuition being that larger models have more parameters with which to fully describe the landscape of volumes occurring in the dataset using the images at hand. We recommend 1024x3 as a baseline at which we expect to be able to characterize the volumes — and the heterogeneity landscape if necessary — in most datasets with a reasonable amount of signal, but of course there may be cases where a larger model is required or a smaller model is sufficient.

In general, it is a good idea to train smaller models on new data to be able to quickly filter out bad particles and perform other sanity checks, and then in the later stages of an experiment to train as large of a model as your compute resources and time constraints will allow. An example of this type of tradeoff in action is Figure 2 from [the original cryoDRGN paper](https://www.nature.com/articles/s41592-020-01049-4), where we see that applying larger models in the case of RAG1-RAG2 and 80S results in a volume more consistent at higher resolutions with a cryoSPARC-reconstructed reference (Fig 2a-b), albeit with an up to 4x longer runtime for 1024x3 compared to 128x3 (Fig 2e):

![image](https://github.com/user-attachments/assets/fba13b59-bb80-47f4-a320-38f93ea44361)

In addition to training with larger models, you may find that the tools available in our analysis notebooks and our `cryodrgn analyze` command are not enough to characterize the landscape of volumes the model found in your dataset, in which case you will also have to try the more computationally intensive tools `cryodrgn analyze_landscape` and `cryodrgn analyze_landscape_full` that directly compare reconstructed volumes to one another, rather than just using their co-ordinates in the latent embedding space to perform clustering!

**Reply by cbeck22** [reply]
That makes sense, thank you! I have a few follow-up questions:

- In which cases would it be better to increase the number of layers vs nodes per layer? For example, the table in Fig 2E lists 1024x3 and 1024x10 as possible architectures
- When would it make sense to change the number of dimensions in the latent space vs changing the size of the neural net? For example, I've found several papers that used a z of 10 instead of the 8 recommended in tutorial.
- In what scenarios would it make sense to use differently sized architectures for the encoder and decoder? For example, I found a paper that used a 1024x3 architecture for the encoder, but a 2048x3 architecture for the decoder. What effect does changing the size of the decoder relative to the encoder have on the types of volumes produced?

**Reply by michal-g** [reply]
It is difficult to give cut-and-dry answers to these questions, as the optimal choice of architecture will often depend on properties specific to a particular dataset. That being said, we can list some general intuitions:

1. Increasing the number of nodes per layer is likely more important than increasing the number of layers; we tend to observe the pattern shown in Fig 2e above, where the marginal gains from tripling the network depth are less than that from doubling the layer size. We suspect that necessary layer count is driven by the general complexity of the task of reconstructing 3D volumes from 2D images, while larger layers help models deal with more noise and more heterogeneity, and thus necessary layer size is more likely to vary across datasets. We thus recommend 128x3 or 256x3 as starting points, and trying e.g. 512x3 before 256x10. One case where the tradeoff may come out more favorably for layer count is if there are very complex heterogeneities in your data.
2. More dimensions in the latent space make it more likely that the model will be able to represent more complex types of heterogeneity (although you may also need to increase the size of the neural net to take advantage of this), so if you expect to see lots of heterogeneity, larger z-dims make sense. The choice between 8 and 10 is more likely driven by the fact that there are extra runtime optimizations available for models whose dimension is a multiple of 8 through [torch.cuda.amp](https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html#), and not everybody chooses or is able to take advantage of these. We recommend 8 as a good starting point, especially if your hardware is compatible with AMP! Note that using more latent dimensions runs the risk of under-fitting, where the model simply tends to put each particle in its own singleton cluster, due to the much greater separability inherent in using more dimensions to represent the latent space; see e.g. [the curse of dimensionality](https://en.wikipedia.org/wiki/Curse_of_dimensionality).
3. There are plenty of cases where a larger encoder could be more important, but we would expect the decoder being larger to be better in more cases, due to it having to reconstruct a 3D volume from a latent representation, an inherently more difficult task than mapping images to the latent representation space. Across your two comments you've mentioned papers trying it both ways; could you share the links to each to make it easier to comment on these particular cases?

One short answer to all three is that you should try as many different architectures as you can and compare the results, using the smaller model → larger models strategy I mentioned earlier, since it’s difficult to know *a priori* what architecture will be the best for a given application.

**Reply by cbeck22** [reply]
Thank you for this detailed response! I agree that it's best to just try different architectures and compare the results, but as our computational resources are a little limited, I'm still trying to develop some intuition on which configurations make the most sense so that I can prioritize them.

I tried searching again for the paper that used differently sized decoders and encoders, but I haven't been able to find it again. I'm not planning on testing this particular configuration, but if I come across it again, I'll share it here.  Thank you again!

Cheers,
cbeck
